src/cache/algorithms/expr_variable_reordering.h

```diff
@@ -0,0 +1,189 @@
+/*******************************************************************\
+ Module: Expressions Variable Reordering
+
+ Author: Rafael Sá Menezes
+
+ Date: April 2020
+\*******************************************************************/
+
+#ifndef ESBMC_EXPR_VARIABLE_REORDERING_H
+#define ESBMC_EXPR_VARIABLE_REORDERING_H
+
+#include <cache/expr_algorithm.h>
+#include <util/irep2_expr.h>
+
+/**
+ *  This class will reorder a expression based on it's naming
+ *  the idea is that a expression in the format:
+ *
+ *  c + a + b + 4 + d == b + a
+ *
+ *  Becomes:
+ *
+ *  a + b + c + d + 4 == a + b
+ *
+ *  Note that that this assumes that a constant propagation algorithm
+ *  was executed, so an expression should have at max 1 constant value.
+ *
+ *  Supported Operators: [+, -, *, /]
+ *  Supported Types (constants/symbols): [Signed Integer, Unsigned Integer]
+ *  Supported Relations: [==, !=, <, >, <=, >=]
+ *
+ *  TODO: Define precedence rules for operators
+ */
+class expr_variable_reordering : public expr_algorithm
+{
+  typedef std::vector<symbol2tc> symbols_vec;
+  typedef std::vector<constant_int2tc> values_vec;
+
+public:
+  explicit expr_variable_reordering(expr2tc &expr) : expr_algorithm(expr)
+  {
+  }
+
+  void run() override;
+
+private:
+  /**
+   * Parse a binary operation and reorders it
+   *
+   * Since some operators have precedence this implementation
+   * should take it into account.
+   *
+   * The reorder will check if LHS and RHS are symbols/values
+   * the precedence will be: x OP y OP z OP value
+   *
+   * TODO: support precedence levels, the current implementation will stop if it finds a different operator
+   *
+   * @param expr binary_operation
+   */
+  void run_on_binop(expr2tc &expr) noexcept;
+
+  /**
+   * Parse a relation by reordering the LHS and RHS
+   *
+   * @param expr relation
+   */
+  void run_on_relation(expr2tc &expr) noexcept;
+
+  /**
+   * Parses and adds the symbol or value
+   *
+   * This receives a reference and generates a copy of it to be later used
+   *
+   * @param op parent of the operation
+   * @param is_lhs truth value to determine which side of the op will be checked
+   * @param symbols
+   * @param values
+   */
+  void add_value(
+    const std::shared_ptr<arith_2ops> op,
+    bool is_lhs,
+    symbols_vec &symbols,
+    values_vec &values);
+
+  /**
+   * Parses and replaces symbols and values
+   *
+   * This works by tacking an expression and manually replacing it's inner
```

Comment:
"... replacing it's inner expressions ..." -> "... replacing its inner expressions ..."